Change preview users to be per course

yscik opened the pull request:
Based on #6288
Fixes #6287 #6284

### Changes proposed in this Pull Request

* Change preview users to be created for the current course only
* Multiple preview users are stored in the teacher's user meta (max one for each course) 
* Only switch to preview user when it's created for the current course

### Testing instructions

* Have multiple courses
* Open one, click 'Preview as Student'
* Confirm only the first course is using the preview user, not the other ones
* Also switch to a preview user for another course, ensure it's a different user

yscik committed:
Change preview users to be per course

Imran92 commented:
The code and the logics look very nice to me, it was a very good decision to make the preview users 'per-course' to avoid some unwanted issues. It's working as expected for me for the Course page in the preview mode. Only when I go in the inner pages, like lessons, it starts using the teacher user instead of switching to the preview user

https://user-images.githubusercontent.com/6820724/207554871-7848324d-918b-45f4-8c07-aa5c8eb9e3c6.mov

yscik committed:
Merge branch 'feature/guest-user' into change/preview-user-course

yscik committed:
Clean up redirect filter in tests

yscik committed:
Fix Sensei_Utils::get_current_course returning string when in a lesson

yscik committed:
Add tests and a few extra checks

yscik committed:
Change preview user numbering

yscik committed:
Fix tearDown signature

yscik commented:
Ah, looks like `get_current_course` returned as string in lessons and as int in courses. Fixed it there in f3f666c, I really hope nothing relies on that behavior :)

Also made a few more small adjustments, and removed `get_user_count` from the preview user creation too. (Since this one is tied to the user and course, I added their IDs instead)

yscik committed:
Fix deleting users on multisite setup

yscik committed:
Adjust user name display in users table

yscik committed:
Spacing

yscik committed:
Merge branch 'feature/guest-user' into change/preview-user-course

Imran92 reviewed (approved):
Works very nicely now 👍 Just a small issue that I've noticed while testing is (which probably is not related to this PR) is if we don't click on 'Take Course' of Open Access course, rather complete a lesson directly, the Take Course button is still visible on the course page. We can create an issue for this and handle it there

yscik commented:
Yes, the new preview student is not yet enrolled — I'll fix that in #6282
